### DISCUSSION TITLE
Can't setup ISO Broadcast Demo on nrf53dk

### DISCUSSION THREAD

**Discussion by kaiser-ren** [zephyrproject-rtos/zephyr#35407] [Q&A]
**Describe the bug**
I compiled and flashed ./samples/bluetooth/iso_broadcast/ on nrf53dk. After power-on the dk, I found that there was an error when host sent HCI_LE_Set_-Extended_-Advertising_Parameters to controller, the controller return an event with status 0x01. 

**To Reproduce**
1. cd ./samples/bluetooth/iso_broadcaster/
2. west build -p auto -b nrf5340dk_nrf5340_cpuapp ./
3. west flash
4. west build -p auto -b nrf5340dk_nrf5340_cpunet ./
5. west flash

**Expected behavior**

**Impact**
Can't setup BIG/BIS

**Logs and console output**
\`\`\`*** Booting Zephyr OS build zephyr-v2.5.0-3928-g86f18a61fa05  ***
Starting ISO Broadcast Demo
Failed to create advertising set (err -5)
[00:00:00.010,437] <inf> bt_hci_core: HW Platform: Nordic Semiconductor (0x0002)
[00:00:00.010,437] <inf> bt_hci_core: HW Variant: nRF53x (0x0003)
[00:00:00.010,437] <inf> bt_hci_core: Firmware: Standard Bluetooth controller (0x00) Version 2.6 Build 0
[00:00:00.011,169] <inf> bt_hci_core: Identity: D5:2B:02:90:B0:4F (random)
[00:00:00.011,169] <inf> bt_hci_core: HCI: version 5.2 (0x0b) revision 0x0000, manufacturer 0x05f1
[00:00:00.011,169] <inf> bt_hci_core: LMP: version 5.2 (0x0b) subver 0xffff
[00:00:00.011,413] <wrn> bt_hci_core: opcode 0x2036 status 0x01
\`\`\

**Environment (please complete the following information):**
 - Commit SHA or Version used: cd9421a45e7254de6be913debe153dbac4be1e91

**Additional context**

**Comment by carlescufi** [comment]
ISO is still very much a work in progress, so we've moved this into a discussion.

@cvinayak @Thalley can you add some feedback here please?

**Reply by Thalley** [reply]
ISO broadcast is working on the host, but has not yet been fully implemented in the controller, although we (or rather Vinayak) is getting close.

**Reply by cvinayak** [reply]
Broadcast ISO is work in progress, implementation for sequential packing is missing integration with ISOAL module which is also work in progress. BIS in controller is here: https://github.com/zephyrproject-rtos/zephyr/pull/31009

**Reply by govindadas** [reply]
What about the "samples/bluetooth/peripheral_iso" (and "/central_iso") example, is this supported by the controller?

**Reply by Thalley** [reply]
@govindadas No, none of the ISO samples are supported by the Zephyr controller at this point.

**Reply by govindadas** [reply]
@Thalley ok, thanks for the clarification!

**Reply by danielhira** [reply]
Hello, I am currently attempting to use the nRF5340 DK (non-audio DK) to run the iso_broadcast_benchmark sample app and was also having issues (with a variety of config options enabled, able to sync between broadcaster + receiver but no packets are received by receiver; unsure if packets are actually sent by the broadcaster).

Is Broadcast ISO in the controller still a work in progress? Or is it supported just on the 5340 audio DK as an experimental feature (https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/applications/nrf5340_audio/README.html)?

Apologies if this is the wrong place to raise the question, I didn't want to open a new discussion since this was still open. In case it helps to determine if my issue warrants a separate discussion, this is my app config (https://pastebin.com/ht7u04C6) and hci_rpmsg config (https://pastebin.com/eKv2rCty), and output from the broadcaster (https://pastebin.com/ipKjyky5) and receiver (https://pastebin.com/yM90kiht). Sequence was: start receiver, start broadcaster, wait ~10s, hit q on the broadcaster. Thanks!

**Reply by Thalley** [reply]
@danielhira Broadcast ISO is still a work in progress and is still experimental. The benchmark sample is a good way to try out different parameters, but at this point probably a lot them won't be supported. 

It can be hard to determine if there is an issue on the broadcast or receiver side when nothing is received, and for something like that you should try to use a sniffer to get an air trace.

**Reply by danielhira** [reply]
Thank you for the confirmation, @Thalley !

**Reply by Thalley** [reply]
If you have any specific parameters that you'd like to use, and they don't seem to work, please let us know :)

**Reply by danielhira** [reply]
@Thalley -- I am just trying to use any form of ISO broadcast to transfer data from one device to several receiver devices, so any parameters that are successful would work great for me :) I assumed the default parameters would be my best bet (as seen by the broadcaster: `(Current settings: rtn=2, interval=7500, latency=10, phy=2, sdu=251, packing=0, framing=0, bis_count=1)`, and as seen by the receiver: `BIGinfo received: num_bis 1, nse 7, interval 7.50 ms, bn 1, pto 1, irc 3, max_pdu 251, sdu_interval 7500 us, max_sdu 251, phy LE 2M, without framing, not encrypted`). I also assumed that I have either mis-configured one or both of the boards, either by over-setting Kconfigs or by missing a Kconfig that needs to be set.

Are there adjustments I should make which are known to be successful at sending packets after the BIG has been created? The only modifications to the iso_broadcast_benchmark app that I've made is setting the various Kconfigs and adding additional debug logging -- I am just trying to get a baseline of sending any data across. I have multiples of the nRF5340 audio, nRF5340 non-audio, and nRF52840 dev boards at my disposal for this. Getting a different board is also potentially an option if you have a recommendation. I am currently in the process of getting a sniffer as well.

**Reply by Thalley** [reply]
@danielhira Could you try the `iso_broadcast` and `iso_receive` samples instead, and see if those work? If those are working, please raise a bug issue in Github, as they should work out-of-the-box. 

If they are working, you could try to replicate their parameters in the benchmark sample - And if it fails in the benchmark sample, but not the other samples, then please also create a Github bug issue for that :)

**Reply by danielhira** [reply]
Hi @Thalley -- apologies for the delayed response. Thanks for pointing to the configs in `iso_broadcast` and `iso_receive`, those helped me craft configs that worked for the 5340 non-audio DK boards for the iso_broadcast_benchmark. I need to do a bit of cleanup to make development better, but for what it's worth these are the configs that work for me:

Receiver: https://pastebin.com/Vph6Av9K [cpu], https://pastebin.com/5vhfYZNe [net]
Broadcaster: https://pastebin.com/Vph6Av9K [cpu], https://pastebin.com/2Yj8RUnq [net]

^ common cpu core config but different net core configs. Keep in mind I'm sure I have unnecessary cruft in there, but that this worked so it's at least a baseline I can iterate on while I optimize :)

Regarding the parameters, I have been successful in sending SDUs up to 246B. 247B errors on the receiver side with `Unexpected ISO end fragment`, and 248-251B appear to send on the broadcaster side but the receiver receives no packets (still working on getting a sniffer, so I unfortunately have no packet captures). Is this expected behavior for SDUs above 246B? I was hoping to use a 251B SDU size. Example parameters: RTN=2, Interval=7500us, Latency=10ms, SDU=244-251, Packing=0, Framing=0, Num BIS=1

I also have thus far been unsuccessful at sending an SDU larger broken across multiple PDUs, example parameters: RTN=2, Interval=10000us, Latency=12ms, SDU=300, Packing=0, Framing=1, Num BIS=1. Is this a known issue? Thanks!

**Reply by Thalley** [reply]
@cvinayak Does the Zephyr controller support sending SDU data when sent over mulitple HCI Packets with the `BT_ISO_START`, `BT_ISO_CONT` and `BT_ISO_END` host->controller segmentation for broadcast?

**Reply by ApplicationNinja** [reply]
Is there any further update here? Do any of the ISO samples work on nRF5340 non-audio dev kit boards?

**Reply by cvinayak** [reply]
> @cvinayak Does the Zephyr controller support sending SDU data when sent over mulitple HCI Packets with the `BT_ISO_START`, `BT_ISO_CONT` and `BT_ISO_END` host->controller segmentation for broadcast?

ISOAL implementation alongwith CI tests is upstream, but I have not tested it with a sample. Contributions covering segmentation use is welcome.

**Reply by cvinayak** [reply]
> Is there any further update here? Do any of the ISO samples work on nRF5340 non-audio dev kit boards?

ISO support in Controller is work in progress. All work happens upstream, basic features are usable with correct use of Kconfig options, and work in progress can be followed using open GitHub issues and pull requests. 
https://github.com/zephyrproject-rtos/zephyr/issues?q=is%3Aopen+is%3Aissue+label%3A%22area%3A+Bluetooth+ISO%22
https://github.com/zephyrproject-rtos/zephyr/pulls?q=is%3Aopen+is%3Apr+label%3A%22area%3A+Bluetooth+ISO%22+

**Reply by ApplicationNinja** [reply]
> > Is there any further update here? Do any of the ISO samples work on nRF5340 non-audio dev kit boards?
> 
> ISO support in Controller is work in progress. All work happens upstream, basic features are usable with correct use of Kconfig options, and work in progress can be followed using open GitHub issues and pull requests. https://github.com/zephyrproject-rtos/zephyr/issues?q=is%3Aopen+is%3Aissue+label%3A%22area%3A+Bluetooth+ISO%22 https://github.com/zephyrproject-rtos/zephyr/pulls?q=is%3Aopen+is%3Apr+label%3A%22area%3A+Bluetooth+ISO%22+

Thank you for your response...

**Comment by kaiser-ren** [comment]
If this sample code is not ready to be used, suggest that not put in ./samples folder because people may try to use it, but the result is the same, failed.

**Reply by Thalley** [reply]
I understand the frustration and/or confusion of this, but at some point it was decided that it was okay to put such features (supported only by the host) in the samples, as those are also useful for the people working on the controller to verify the implementation and behavior (see also https://github.com/zephyrproject-rtos/zephyr/issues/32669#issuecomment-786087653)

Perhaps we need to find a way to let it be clear when something runs on Zephyr (full-stack) or only on Zephyr host. This is actually similar to the PR (https://github.com/zephyrproject-rtos/zephyr/pull/33579) I made for one of your other Github issues (https://github.com/zephyrproject-rtos/zephyr/issues/32669).

**Reply by kaiser-ren** [reply]
> I understand the frustration and/or confusion of this, but at some point it was decided that it was okay to put such features (supported only by the host) in the samples, as those are also useful for the people working on the controller to verify the implementation and behavior (see also [#32669 (comment)](https://github.com/zephyrproject-rtos/zephyr/issues/32669#issuecomment-786087653))
> 
> Perhaps we need to find a way to let it be clear when something runs on Zephyr (full-stack) or only on Zephyr host. This is actually similar to the PR (#33579) I made for one of your other Github issues (#32669).

That is so kind. I'm sorry that I raised this issue twice. 

IMHO, separated full-stack, controller and host will be a great option to make developers select what they want to try in Bluetooth samples, for example, below or above HCI. There are so many boards/platforms/chipsets, and a lot of more are upcoming, different boards/platforms/chipsets have different capabilities.
